instrumentation/play/play-ws/play-ws-common/testing/src/main/java/io/opentelemetry/javaagent/instrumentation/playws/PlayWsClientBaseTest.java

```diff
@@ -0,0 +1,159 @@
+import io.opentelemetry.semconv.ServerAttributes;
+import java.io.IOException;
+import java.net.InetAddress;
+import java.net.URI;
+import java.net.UnknownHostException;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.extension.RegisterExtension;
+import play.shaded.ahc.io.netty.resolver.InetNameResolver;
+import play.shaded.ahc.io.netty.util.concurrent.EventExecutor;
+import play.shaded.ahc.io.netty.util.concurrent.ImmediateEventExecutor;
+import play.shaded.ahc.io.netty.util.concurrent.Promise;
+import play.shaded.ahc.org.asynchttpclient.AsyncHttpClient;
+import play.shaded.ahc.org.asynchttpclient.AsyncHttpClientConfig;
+import play.shaded.ahc.org.asynchttpclient.DefaultAsyncHttpClient;
+import play.shaded.ahc.org.asynchttpclient.DefaultAsyncHttpClientConfig;
+import play.shaded.ahc.org.asynchttpclient.RequestBuilderBase;
+
+abstract class PlayWsClientBaseTest<REQUEST> extends AbstractHttpClientTest<REQUEST> {
+
+  @RegisterExtension
+  static final InstrumentationExtension testing = HttpClientInstrumentationExtension.forAgent();
+
+  static final AutoCleanupExtension autoCleanup = AutoCleanupExtension.create();
+
+  private static ActorSystem system;
+  protected static AsyncHttpClient asyncHttpClient;
+  protected static AsyncHttpClient asyncHttpClientWithReadTimeout;
+  protected static ActorMaterializer materializer;
+
+  @BeforeEach
+  void setup() {
+    String name = "play-ws";
+    system = ActorSystem.create(name);
+    materializer = ActorMaterializer.create(ActorMaterializerSettings.create(system), system, name);
+
+    // Replace dns name resolver with custom implementation that returns only once address for each
+    // host. This is needed for "connection error dropped request" because in case of connection
+    // failure ahc will try the next address which isn't necessary for this test.
+    RequestBuilderBase.DEFAULT_NAME_RESOLVER =
+        new CustomNameResolver(ImmediateEventExecutor.INSTANCE);
+
+    asyncHttpClient = createClient(false);
+    asyncHttpClientWithReadTimeout = createClient(true);
+    autoCleanup.deferCleanup(asyncHttpClient);
+    autoCleanup.deferCleanup(asyncHttpClientWithReadTimeout);
+  }
+
+  @AfterEach
+  void tearDown() throws IOException {
+    if (system != null) {
+      system.terminate();
+    }
+    if (asyncHttpClient != null) {
+      asyncHttpClient.close();
+    }
+    if (asyncHttpClientWithReadTimeout != null) {
+      asyncHttpClientWithReadTimeout.close();
+    }
+  }
+
+  @Override
+  public REQUEST buildRequest(String method, URI uri, Map<String, String> headers)
+      throws Exception {
+    return null;
+  }
+
+  @Override
+  public int sendRequest(REQUEST request, String method, URI uri, Map<String, String> headers)
+      throws Exception {
+    return 0;
+  }
```

Comment:
Why do you need these here? Don't you need to provide implementation for these in a subclass anyway?

Reply:
original groovy class had it. and play-ws-common has it too. I can remove them in both places and make sure tests are passing.